### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by lucas-crodrigues**
This is part 3 of the School Library app initiated in the third week, of the fourth module in the Microverse program.

The requirements are:

> ### Project requirements
> - Create a class `Classroom` with the following:
>   - `@label` instance variable, should be initialized in the constructor.
>   - Setter and getter for `@label` (remember about `attr_accessor`).
> - Create the `has-many`/`belongs-to` relationship between `Classroom` and `Student`. The following should be implemented:
>   - Create the `has-many` side (a classroom has many students).
>   - Create the `belongs-to` side (a student belongs to a classroom).
>   - Make sure that when adding a student to a classroom it also sets the classroom for the student.
>   - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
> - Create a class `Book` with the following:
>   - `@title` and `@author` instance variables, should be initialized in the constructor.
>   - Setters and getters for instance variables (remember about `attr_accessor`).
> - Create a class `Rental` with the following:
>   - `@date` instance variable, should be initialized in the constructor.
>   - Setter and getter for `@date` (remember about `attr_accessor`).
> - Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following should be implemented:
>   - Create the `has-many` side of `Book` and `Rental` (a book has many rentals).
>   - Create the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
>   - Create the `has-many` side of `Person` and `Rental` (a person has many rentals).
>   - Create the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
>   - Modify the constructor of `Rental` so `Book` and `Person` are set in it.

- [x] [Create a class 'Classroom'](https://github.com/lucas-crodrigues/school-library/commit/ae43944d3bf4ffaa0d790284ca7265ee2c2c2b21)
- [x] [Create a has-many/belongs-to relationship between 'Classroom' and 'Student'](https://github.com/lucas-crodrigues/school-library/commit/be48173323fa4c75f0eaf9d23911423a3754dbd2)
- [x] [Create a class 'Book'](https://github.com/lucas-crodrigues/school-library/commit/8b52122adc99b4d09e3d3ff4bce4a5af2030abce)
- [x] [Create a class 'Rental'](https://github.com/lucas-crodrigues/school-library/commit/e13c306362c9802d469c791f99517fcb0b1844a7)
- [x] [Create the many-to-many relationship between 'Person' and 'Book' using the intermediate class 'Rental'](https://github.com/lucas-crodrigues/school-library/commit/2f915d472d2d928c9918d5c8adddfafb1a7e36eb)

Thank you.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (FrederickMih) [inline book.rb]: - [ ] I think the rentals variable can also be modified: `attr_accessor :title, :author, :rentals`. Let's fix this.

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals.push(rental)
+    rental.book = self
+  end
```
Comment (FrederickMih) [inline book.rb]: - [ ] We shall rent out a book to a particular `person`(teacher or student) on a particular date. What do you think if we implement the method like this?:

```ruby
  def add_rental(person, date)
    # *the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 
     Rental.new(date, self, person)
  end
```

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (FrederickMih) [inline rental.rb]: - [ ] These variables are not only accessible: They can equally be modified: `attr_accessor :date, :book, :person`

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+  def initialize(date, person, book)
```
Comment (FrederickMih) [inline rental.rb:4]: - Fantastic job for breaking the `many-to-many relationship` by setting the Book and Person into this constructor :+1:

File: person.rb [hunk 5]
```diff
@@ -4,14 +4,15 @@
 
 class Person < Nameable
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (FrederickMih) [inline person.rb]: - [ ] As already mentioned above, the rental variable can equally be modified. Let's kindly make it an `attr_accessor`

File: person.rb [hunk 6]
```diff
@@ -22,6 +23,11 @@ def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @rentals.push(rental)
+    rental.person = self
+  end
```
Comment (FrederickMih) [inline person.rb]: - [Optional] We can also have something like this:

```ruby
  def add_rental(book, date)
     # *The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association
      Rental.new(date, book, self)
  end
```

**Review by FrederickMih — CHANGES_REQUESTED** [review]
Hi @lucas-crodrigues, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :green_circle: 

- Pull Request has a great title and summary :+1: 
- No Linter issue detected :+1: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 1 above)
Comment (lucas-crodrigues) [reply book.rb]: Will do, thank you

File: book.rb (hunk 2 above)
Comment (lucas-crodrigues) [reply book.rb]: Will do, thank you

File: rental.rb (hunk 3 above)
Comment (lucas-crodrigues) [reply rental.rb]: Will do, thank you

File: rental.rb (hunk 4 above)
Comment (lucas-crodrigues) [reply rental.rb:4]: Thank you

File: person.rb (hunk 5 above)
Comment (lucas-crodrigues) [reply person.rb]: Will do, thank you

File: person.rb (hunk 6 above)
Comment (lucas-crodrigues) [reply person.rb]: Will do, thank you

File: book.rb (hunk 1 above)
Comment (ShoiraTa) [reply book.rb]: Resolved 👍🏻

**Review by ShoiraTa — APPROVED** [review]
## **STATUS: APPROVED  🟢 ** 
------
### Hi @lucas-crodrigues , 

**Great work on making the changes requested by a previous reviewer 👏🏻** 

### Highlights:
 ✔️ Project is setup correctly
 ✔️ All linters are passing
✔️ Correct gitflow
 ✔️ Professional documentation

Your project is complete! . There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

image

**Cheers and Happy coding! :clap: :clap: :clap:**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [@shoirata] in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
***
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have  limited number of  reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**